File: RailgunVB/Commands/Fun/Rst.vb
Imports System.IO
Imports System.Text
Imports Discord
Imports Discord.Commands
Imports RailgunVB.Core
Imports RailgunVB.Core.Configuration
Imports RailgunVB.Core.Preconditions
Imports TreeDiagram
Imports TreeDiagram.Models.Server.Fun

Namespace Commands.Fun
    
    <Group("rst")>
    Public Class Rst
        Inherits SystemBase
    
        Private ReadOnly _config As MasterConfig

        Public Sub New(config As MasterConfig)
            _config = config
        End Sub
        
        <Command>
        Public Async Function RstAsync() As Task
            Dim data As FunRst = Await Context.Database.FunRsts.GetAsync(Context.Guild.Id)
            
            If data Is Nothing
                await ReplyAsync($"RST is empty! Please add some stuff using {Format.Code(
                    $"{_config.DiscordConfig.Prefix}rst add [message]")}.")
                Return
            ElseIf Not (data.IsEnabled)
                await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.")
                Return
            End If
            
            Dim msg As String = data.GetRst()
            
            If String.IsNullOrEmpty(msg) Then msg = $"RST is empty! Please add some stuff using {Format.Code(
                $"{_config.DiscordConfig.Prefix}rst add [message]")}."
            
            await ReplyAsync(msg)
        End Function
        
        <Command("add")>
        Public Async Function AddAsync(<Remainder> msg As String) As Task
            If String.IsNullOrWhiteSpace(msg)
                Await ReplyAsync("Your message was empty. Please add a message to add.")
                Return
            End If
            
            Dim data As FunRst = Await Context.Database.FunRsts.GetOrCreateAsync(Context.Guild.Id)
            
            If Not (data.IsEnabled)
                Await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.")
                Return
            End If
            
            data.AddRst(msg)
            
            Await ReplyAsync($"Added To RST: {Format.Code(msg)}")
        End Function
        
        <Command("remove"), UserPerms(GuildPermission.ManageMessages)>
        Public Async Function RemoveAsync(index As Integer) As Task
            Dim data As FunRst = Await Context.Database.FunRsts.GetAsync(Context.Guild.Id)
            
            If data Is Nothing
                await ReplyAsync($"RST is empty! Please add some stuff using {Format.Code(
                    $"{_config.DiscordConfig.Prefix}rst add [message]")}.")
                Return
            ElseIf Not (data.IsEnabled)
                await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.")
                Return
            ElseIf index < 0 OrElse index >= data.Rst.Count
                await ReplyAsync("The Message Id provided is out of bounds. Please recheck via RST List.")
                Return
            End If
            
            data.Rst.RemoveAt(index)
            
            await ReplyAsync("Message Removed!")
        End Function
        
        <Command("list"), BotPerms(ChannelPermission.AttachFiles)>
        Public Async Function ListAsync() As Task
            Dim data As FunRst = Await Context.Database.FunRsts.GetAsync(Context.Guild.Id)
            
            If data Is Nothing
                await ReplyAsync($"RST is empty! Please add some stuff using {Format.Code(
                    $"{_config.DiscordConfig.Prefix}rst add [message]")}.")
                Return
            ElseIf Not (data.IsEnabled)
                await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.")
                Return
            End If
            
            Dim output As New StringBuilder
            
            output.AppendLine("Randomly Selected Text List :").AppendLine()
            
            For Each msg As String In data.Rst
                output.AppendFormat("[{0}] {1}", Format.Code(data.Rst.IndexOf(msg).ToString()), msg).AppendLine()
            Next
            
            If output.Length < 1950
                await ReplyAsync(output.ToString())
                Return
            End If
            
            Dim filename As String = $"{Context.Guild.Id}_RST.txt"
            
            Await File.WriteAllTextAsync(filename, output.ToString())
            await Context.Channel.SendFileAsync(filename)
            
            File.Delete(filename)
        End Function
        
        <Command("allowdeny"), UserPerms(GuildPermission.ManageMessages)>
        Public Async Function AllowDenyAsync() As Task
            Dim data As FunRst = Await Context.Database.FunRsts.GetOrCreateAsync(Context.Guild.Id)
            
            data.IsEnabled = Not (data.IsEnabled)
            
            await ReplyAsync($"RST is now {If(data.IsEnabled, Format.Bold("enabled"), Format.Bold("disabled"))}!")
        End Function
        
        <Command("reset"), UserPerms(GuildPermission.ManageMessages)>
        Public Async Function ResetAsync() As Task
            Dim data As FunRst = Await Context.Database.FunRsts.GetAsync(Context.Guild.Id)
            
            If data Is Nothing
                await ReplyAsync("RST has no data to reset.")
                Return
            End If
            
            Context.Database.FunRsts.Remove(data)
            
            await ReplyAsync("RST has been reset.")
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Commands/Fun/Rst.cs
using System.Text;
using System.Threading.Tasks;
using Discord;
using Finite.Commands;
using Railgun.Core.Commands;
using Railgun.Core.Commands.Attributes;
using Railgun.Core.Configuration;
using Railgun.Core.Utilities;
using TreeDiagram;

namespace Railgun.Commands.Fun
{
	[Alias("rst")]
	public class Rst : SystemBase
	{
		private readonly MasterConfig _config;

		public Rst(MasterConfig config) => _config = config;

		[Command]
		public async Task RstAsync()
		{
			var data = Context.Database.FunRsts.GetData(Context.Guild.Id);

			if (data == null) {
				await ReplyAsync($"RST is empty! Please add some stuff using {Format.Code($"{_config.DiscordConfig.Prefix}rst add [message]")}.");

				return;
			} else if (!data.IsEnabled) {
				await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.");

				return;
			}

			var msg = data.GetRst();

			if (string.IsNullOrEmpty(msg))
				msg = $"RST is empty! Please add some stuff using {Format.Code($"{_config.DiscordConfig.Prefix}rst add [message]")}.";

			await ReplyAsync(msg);
		}

		[Command("add")]
		public async Task AddAsync([Remainder] string msg)
		{
			if (string.IsNullOrWhiteSpace(msg)) {
				await ReplyAsync("Your message was empty. Please add a message to add.");

				return;
			}

			var data = Context.Database.FunRsts.GetOrCreateData(Context.Guild.Id);

			if (!data.IsEnabled) {
				await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.");

				return;
			}

			data.AddRst(msg);

			await ReplyAsync($"Added To RST: {Format.Code(msg)}");
		}

		[Command("remove"), UserPerms(GuildPermission.ManageMessages)]
		public async Task RemoveAsync(int index)
		{
			var data = Context.Database.FunRsts.GetData(Context.Guild.Id);

			if (data == null) {
				await ReplyAsync($"RST is empty! Please add some stuff using {Format.Code($"{_config.DiscordConfig.Prefix}rst add [message]")}.");

				return;
			} else if (!data.IsEnabled) {
				await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.");

				return;
			} else if (index < 0 || index >= data.Rst.Count) {
				await ReplyAsync("The Message Id provided is out of bounds. Please recheck via RST List.");

				return;
			}

			data.Rst.RemoveAt(index);

			await ReplyAsync("Message Removed!");
		}

		[Command("list"), BotPerms(ChannelPermission.AttachFiles)]
		public async Task ListAsync()
		{
			var data = Context.Database.FunRsts.GetData(Context.Guild.Id);

			if (data == null) {
				await ReplyAsync($"RST is empty! Please add some stuff using {Format.Code($"{_config.DiscordConfig.Prefix}rst add [message]")}.");

				return;
			} else if (!data.IsEnabled) {
				await ReplyAsync($"RST is currently {Format.Bold("disabled")} on this server.");

				return;
			}

			var output = new StringBuilder()
				.AppendLine("Randomly Selected Text List :").AppendLine();

			data.Rst.ForEach(msg => output.AppendFormat("[{0}] {1}", Format.Code(data.Rst.IndexOf(msg).ToString()), msg).AppendLine());

			if (output.Length < 1950) {
				await ReplyAsync(output.ToString());

				return;
			}

			await CommandUtils.SendStringAsFileAsync((ITextChannel)Context.Channel, "RST.txt", output.ToString());
		}

		[Command("allowdeny"), UserPerms(GuildPermission.ManageMessages)]
		public async Task AllowDenyAsync()
		{
			var data = Context.Database.FunRsts.GetOrCreateData(Context.Guild.Id);

			data.IsEnabled = !data.IsEnabled;

			await ReplyAsync($"RST is now {(data.IsEnabled ? Format.Bold("enabled") : Format.Bold("disabled"))}!");
		}

		[Command("reset"), UserPerms(GuildPermission.ManageMessages)]
		public async Task ResetAsync()
		{
			var data = Context.Database.FunRsts.GetData(Context.Guild.Id);

			if (data == null) {
				await ReplyAsync("RST has no data to reset.");

				return;
			}

			Context.Database.FunRsts.Remove(data);

			await ReplyAsync("RST has been reset.");
		}
	}
}